src/smp.rs

```diff
@@ -43,54 +46,119 @@ impl PerCpu {
+                // important for the stack (in the boot CPU case, pages are dirty because the stack is
+                // in use).
+                // Safe because the chunk is composed by hypervisor pages owned by `pcpu_stack`.
+                unsafe {
+                    SequentialPages::<InternalDirty>::from_mem_range(
+                        pcpu_stack.base(),
+                        pcpu_stack.page_size(),
+                        pcpu_stack.len(),
+                    )
+                    .unwrap()
+                }
+            };
+            let stack_top_addr = stack_pages
+                .base()
+                .checked_add_pages(stack_pages.len())
+                .unwrap();
             let ptr = Self::ptr_for_cpu(cpu_id);
             let pcpu = PerCpu {
                 cpu_id,
                 vmid_tracker: RefCell::new(VmIdTracker::new()),
-                page_table: HypMap::get().new_page_table(hyp_mem),
+                page_table: HypMap::get().new_page_table(hyp_mem, stack_pages),
                 umode_task: Once::new(),
                 online: Once::new(),
+                stack_top_addr,
             };
             // Safety: ptr is guaranteed to be properly aligned and point to valid memory owned by
             // PerCpu. No other CPUs are alive at this point, so it cannot be concurrently modified
             // either.
             unsafe { core::ptr::write(ptr as *mut PerCpu, pcpu) };
         }
 
-        // Load TP with the address of our PerCpu struct so that we're consistent with secondary
-        // CPUs once they're brought up.
-        let my_tp = Self::ptr_for_cpu(cpu_info.hart_id_to_cpu(boot_hart_id as u32).unwrap()) as u64;
+        // Initialize TP register and set this CPU online to be consistent with secondary CPUs.
+        Self::setup_this_cpu(boot_hart_id);
+        let me = Self::this_cpu();
+        me.set_online();
+    }
+
+    /// Initializes the TP pointer to point to PerCpu data.
+    pub fn setup_this_cpu(hart_id: u64) {
+        let cpu_info = CpuInfo::get();
+        let cpu_id = cpu_info
+            .hart_id_to_cpu(hart_id as u32)
+            .expect("Cannot find CPU ID");
+
+        // Load TP with the address of our PerCpu struct.
+        let tp = Self::ptr_for_cpu(cpu_id) as u64;
         unsafe {
             // Safe since we're the only users of TP.
-            asm!("mv tp, {rs}", rs = in(reg) my_tp)
+            asm!("mv tp, {rs}", rs = in(reg) tp)
         };
+    }
 
-        let me = Self::this_cpu();
-        me.set_online();
+    fn boot_cpu_stack() -> SequentialPages<InternalDirty> {
+        // Get the pages of the current stack as created by the linker.
+        // Safe because these are linker created variables.
+        let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+        let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+        let stack_startaddr = PageAddr::new(RawAddr::supervisor(stack_start))
```

Comment:
In cases like this I think variable shadowing is fine it would improve the readability.

Reply:
Readability is subjective. :-)

I find it more readable to step things in small steps. But what you're suggesting is more in line with the rust way so I'll happily change that.

Reply:
Changed. Put unsafe across the pageaddr block. Let me know what you think.